File: internal/pkg/ratelimitx/scripts/token_bucket.lua
local key = KEYS[1]
local now = tonumber(ARGV[1])
local window = tonumber(ARGV[2])
local limit = tonumber(ARGV[3])
local capacity = limit

local tokens = tonumber(redis.call("HGET", key, "tokens"))
local ts = tonumber(redis.call("HGET", key, "ts"))

if tokens == nil then
    tokens = capacity
end
if ts == nil then
    ts = now
end

local elapsed = now - ts
if elapsed < 0 then
    elapsed = 0
end

tokens = math.min(capacity, tokens + elapsed * capacity / window)

local allowed = 0
if tokens >= 1 then
    tokens = tokens - 1
    allowed = 1
end

redis.call("HSET", key, "tokens", tokens, "ts", now)
redis.call("PEXPIRE", key, window * 2)

return {allowed, math.floor(tokens)}

File: internal/pkg/ratelimitx/token_bucket.go
package ratelimitx

import (
	"math"
	"time"
)

type memTokenBucket struct {
	tokens float64
	last   time.Time
}

func (m *memLimiter) allowTokenBucket(key string, limit int64, window time.Duration, now time.Time) (bool, int64) {
	capacity := float64(limit)
	item, ok := m.tokens[key]
	if !ok {
		item = memTokenBucket{
			tokens: capacity,
			last:   now,
		}
	} else {
		elapsed := now.Sub(item.last)
		if elapsed < 0 {
			elapsed = 0
		}
		item.tokens = math.Min(capacity, item.tokens+elapsed.Seconds()*capacity/window.Seconds())
		item.last = now
	}

	if item.tokens < 1 {
		m.tokens[key] = item
		return false, floorRemaining(item.tokens)
	}
	item.tokens--
	m.tokens[key] = item
	return true, floorRemaining(item.tokens)
}
